### PR TITLE
Add ERC20 and ERC721 warnings for withdrawl

### PR DESCRIPTION
**Pull request by alexkeating**
### Issue
#310 

### Description

Added a check for an ERC721 and ERC20

[Screencast from 10-02-2022 11:58:42 PM.webm](https://user-images.githubusercontent.com/12705146/193499324-22e957f6-9d09-4d2a-9c86-4394953852fc.webm)

### CODE REVIEW

File: frontend/src/i18n/locales/zh-cn.json
Comment (alexkeating) [inline frontend/src/i18n/locales/zh-cn.json:310]: @garyghayrat Are these translations okay?

Comment (garyghayrat) [reply frontend/src/i18n/locales/zh-cn.json:310]: Yes!

Comment (mds1) [reply frontend/src/i18n/locales/zh-cn.json:310]: One nit is to just add hyphens in both i18n files, e.g. "ERC-20" and "ERC-721"

File: frontend/src/utils/address.ts
Comment (mds1) [inline frontend/src/utils/address.ts:143]: Should we just push one `'Utils.Address.is-erc721'` string instead of composing? Seems that would scale better

File: frontend/src/utils/utils.ts
Comment (mds1) [inline frontend/src/utils/utils.ts:159]: We probably should look for a `tokenURI` to identify an NFT. I've seen contracts that have a `name` and `symbol` for other reasons, but aren't tokens

Comment (alexkeating) [reply frontend/src/utils/utils.ts:159]: That is a good point. The tricky thing with tokenURI is it requires an id argument. If we include it should we pass in an id of 1?

File: frontend/src/utils/address.ts
Comment (alexkeating) [reply frontend/src/utils/address.ts:143]: I am changing to a single string, but this construction matches the way existing error messages are constructed for withdrawl

Comment (mds1) [reply frontend/src/utils/address.ts]: Ah, ok yea we might need to do some refactoring then if we ever support more languages, but thanks!

File: frontend/src/utils/utils.ts
Comment (mds1) [reply frontend/src/utils/utils.ts:159]: Ahh, also a good point. Hm ok so wdyt of removing the ERC-20/721 distinction, looking for name and symbol, and if both exist say something like "this might be a token contract, such as ERC-20 or ERC-721". That way it's simpler, and other contracts that have name/symbol prob aren't ones you want to withdraw to anyway

Comment (alexkeating) [reply frontend/src/utils/utils.ts:159]: I think that is a great idea, I will remove the distinction.

**Comment by alexkeating** [comment]
Removed distinction

File: frontend/src/utils/address.ts
Comment (mds1) [inline frontend/src/utils/address.ts:135]: Nit
```suggestion
  const destIsToken = await isToken(destinationAddress, provider);
```

File: frontend/src/utils/constants.ts
Comment (mds1) [inline frontend/src/utils/constants.ts:22]: We already have `ERC20_ABI` so we don't need to add `TOKEN_ABI`

**Review by mds1 — APPROVED** [review]
Two nits, feel free to self-resolve once addressed, otherwise LGTM. Pretty sure I tested this the other day and it worked as expected, so will wait for @apbendi as second reviewer before merging

File: frontend/src/utils/address.ts
Comment (alexkeating) [reply frontend/src/utils/address.ts:135]: Good catch

File: frontend/src/utils/constants.ts
Comment (alexkeating) [reply frontend/src/utils/constants.ts:22]: True, I had that originally, but thought it might be clearer to separate out into a separate ABI even though it is captured by the ERC20. I will remove and change to the ERC-20 abi

**Review by apbendi — COMMENTED** [review]
Great work @alexkeating, this is working great. Here's my only thought after testing it out: I feel like including it as a normal warning does not give sufficient gravity to what's actually happening. It is very likely that the user doesn't mean to do this, and will lose funds if he does.

Given that, I'm wondering: how hard would it be to show a *different* message if this criteria is met? This would supersede the other warning modal. If this is easy enough, let's do it in this PR. If it's not easy, let's pull it into an issue and merge this, as it's already an improvement.

The copy for the alternate modal could be something like:

### 🚨 Danger

It looks like you're trying to withdraw your funds to token contract. It is very likely this is not what you intend, and proceeding will likely result in a loss of funds. **Do not proceed unless you know exactly what you are doing.**

☐ I Know What I'm Doing

[ Withdraw Anyway ]

**Comment by mds1** [comment]
That's a good point, all other checks are privacy related whereas this would be lost funds. FWIW this is how we spec'd it out in #310, so my suggestion would be to merge as-is since it's better than nothing like we currently have, then we can open a separate issue/PR to make this more prominent

File: frontend/src/i18n/locales/zh-cn.json
Comment (garyghayrat) [reply frontend/src/i18n/locales/zh-cn.json:303]: yes!

**Comment by alexkeating** [comment]
This should be good, I also need to add the joi fix after rebaseing

File: frontend/package.json
Comment (mds1) [inline frontend/package.json:27]: _(arbitrary comment location to start a thread)_

Seems that when I cancel a withdraw to a token after the warning, the withdraw form is stuck in this loading state

image

File: frontend/src/i18n/locales/en-us.json
Comment (mds1) [inline frontend/src/i18n/locales/en-us.json:253]:
```suggestion
    "loss-warning": "It looks like you're trying to withdraw your funds to a token contract. It is very likely this is not what you intend, and proceeding will likely result in a loss of funds. Do not proceed unless you know exactly what you are doing.",
```

Comment (alexkeating) [reply frontend/src/i18n/locales/en-us.json]: This should be resolved

File: frontend/package.json
Comment (alexkeating) [reply frontend/package.json:27]: This should be resolved

**Comment by garyghayrat** [comment]
Getting this error when trying to connect wallet not sure why
image

**Review by garyghayrat — APPROVED** [review]
Two small nits, otherwise working great. Thank you!

**Review by apbendi — COMMENTED** [review]
Hey @alexkeating, sorry for one more nit, but here's something I noticed in testing 🙈 

If the user clicks the checkbox and does the approval, the modal does not dismiss after the withdrawal. The warning modal, by comparison, dismisses before showing the subsequent fee modal.

**Comment by alexkeating** [comment]
> Hey @alexkeating, sorry for one more nit, but here's something I noticed in testing see_no_evil
> 
> If the user clicks the checkbox and does the approval, the modal does not dismiss after the withdrawal. The warning modal, by comparison, dismisses before showing the subsequent fee modal.

Good catch pushed up a fix

**Review by apbendi — APPROVED** [review]
Looks and works great! Thanks for bearing with us and ironing out the details @alexkeating.

**Pull request merged**
